ceno_rt/src/mmio.rs

```diff
@@ -51,3 +51,51 @@ where
 {
     rkyv::access::<T, Failure>(read_slice()).expect("Deserialised access failed.")
 }
+
+/// The memory region with public io.
+fn pubio_region<'a>() -> &'a [u8] {
+    extern "C" {
+        /// The address of this variable is the start of the hints ROM.
+        ///
+        /// It is defined in the linker script.  The value of this variable is undefined.
+        static _pubio_start: u8;
+        /// The _address_ of this variable is the length of the hints ROM
+        ///
+        /// It is defined in the linker script.  The value of this variable is undefined.
+        static _pubio_length: usize;
+    }
+    unsafe {
+        let pubio_length = &_pubio_length as *const usize as usize;
+        from_raw_parts(&_pubio_start, pubio_length)
+    }
+}
+
+/// Get the length of the next pubio
+fn pubio_len() -> usize {
+    extern "C" {
+        /// The address of this variable is the start of the slice that holds the length of the hints.
+        ///
+        /// It is defined in the linker script.  The value of this variable is undefined.
+        static _lengths_of_pubio_start: usize;
+    }
+    static mut NEXT_PUBIO_LEN_AT: *const usize = &raw const _lengths_of_pubio_start;
+    unsafe {
+        let len: usize = core::ptr::read(NEXT_PUBIO_LEN_AT);
+        NEXT_PUBIO_LEN_AT = NEXT_PUBIO_LEN_AT.add(1);
+        len
+    }
+}
+
+pub fn pubio_read_slice<'a>() -> &'a [u8] {
+    &pubio_region()[..pubio_len()]
```

Comment:
How is this mechanism intended to work?

As it stands each call to `pubio_read_slice` returns a slice starting at the same address, but with a different size, following the sequence in `_lengths_of_pubio_start` above.

Reply:
This follow by hints IO. I found it's very tricky here due to the right alignment of `rkyv`, and better having a clear documentation.

So the memory layout for pub io will be like
```
PubIOMem = [end_addr1, end_addr2, end_addr3, ....end_addrN, <rkyv serialized data 1>,  <rkyv serialized data 2>, ... <rkyv serialized data N>]
```
Where each end_addr point to the end address of each serialized data in one to one manner.

`_lengths_of_pubio_start` is a mutable pointer point, per public io read it will return next end_addr, then rkyv::access(PubIOMem[0....END]) will serialize those part of data from the end to start

Reply:
Got it, thanks.